plenum/server/catchup/node_leecher_service.py

```diff
@@ -0,0 +1,125 @@
+
+AllLedgersCaughtUp = NamedTuple('AllLedgerCaughtUp', [])
+
+
+class NodeLeecherService:
+    class State(Enum):
+        Idle = 0
+        SyncingAudit = 1
+        SyncingPool = 2
+        SyncingOthers = 3
+
+    def __init__(self,
+                 config: object,
+                 input: RxChannel,
+                 output: TxChannel,
+                 timer: TimerService,
+                 metrics: MetricsCollector,
+                 provider: CatchupDataProvider):
+        self._config = config
+        self._input = input
+        self._output = output
+        self._timer = timer
+        self.metrics = metrics
+        self._provider = provider
+
+        self._state = self.State.Idle
+
+        # TODO: Get rid of this, theoretically most ledgers can be synced in parallel
+        self._current_ledger = None  # type: Optional[int]
+
+        self._leecher_outbox, rx = create_direct_channel()
+
+        self._leechers = {}  # type: Dict[int, LedgerLeecherService]
+
+    def __repr__(self):
+        return "{}:NodeLeecherService".format(self._provider.node_name())
+
+    def register_ledger(self, ledger_id: int):
+        self._leechers[ledger_id] = \
+            LedgerLeecherService(ledger_id=ledger_id,
+                                 config=self._config,
+                                 input=self._input,
+                                 output=self._leecher_outbox,
+                                 timer=self._timer,
+                                 metrics=self.metrics,
+                                 provider=self._provider)
+
+    def start(self, request_ledger_statuses: bool):
+        for leecher in self._leechers.values():
+            leecher.reset()
+
+        self._state = self.State.SyncingAudit
+        self._leechers[AUDIT_LEDGER_ID].start(request_ledger_statuses)
+
+    def _on_ledger_leecher_stop(self, msg: LedgerCatchupComplete):
+        if self._state == self.State.SyncingAudit:
+            self._on_audit_synced(msg)
+        elif self._state == self.State.SyncingPool:
+            self._on_pool_synced(msg)
+        elif self._state == self.State.SyncingOthers:
+            self._on_other_synced(msg)
+        else:
+            logger.warning("{} got unexpected catchup complete {} during idle state".format(self, msg))
+
+    def _on_audit_synced(self, msg):
+        if msg.ledger_id != AUDIT_LEDGER_ID:
+            logger.warning("{} got unexpected catchup complete {} during syncing audit ledger".format(self, msg))
+            return
+
+        self._state = self.State.SyncingPool
+        self._leechers[POOL_LEDGER_ID].start(request_ledger_statuses=True)
+
+    def _on_pool_synced(self, msg):
+        if msg.ledger_id != POOL_LEDGER_ID:
+            logger.warning("{} got unexpected catchup complete {} during syncing pool ledger".format(self, msg))
+            return
+
+        self._state = self.State.SyncingOthers
+        self._sync_next_ledger()
+
+    def _on_other_synced(self, msg):
```

Comment:
Do we really need 3 on_synced methods?
I think we could create just one generic method (like the current `_on_other_synced`).
The only change is to set the state according to `msg.ledger_id`.

Reply:
I agree that this looks like some code duplication, however I wanted to make state machine as explicit as possible. I'll think about improving this.